proposals/runtime_on_load_on_enabled_events.md

```diff
@@ -0,0 +1,181 @@
+profile first starts. The listener will be given the cause so they can decide to
+take different actions based on the reason it was added to the set. There is
+currently no generic event that encapsulates all instances of extension install
+(and update), startup, and being enabled from a disabled state.
+
+#### Use Cases `chrome.runtime.onEnabled`
+A developer wants to run code once in an extension’s "lifecycle", without storing
+a flag to keep track of this.
+
+Before this API the developer would need to manually keep track of this state
+via a storage flag in persistent storage.
+
+With this new API a single listener can be registered and reduce the extension’s
+complexity.
+
+More use cases can be found [here](https://github.com/w3c/webextensions/issues/353#issuecomment-1490078217).
+
+#### Use Cases `chrome.runtime.onExtensionLoaded`
+An extension automatically does initial bootstrapping, such as creating an
+initial state used throughout the browser session, creating a WebSocket
+connection, checking for things you've missed while the browser was closed, and
+putting something in chrome.storage.session. 
+
+Before this API the developer would have to create multiple listeners for
+`chrome.runtime.onInstalled()`, `onStartup()`, and (above) `onEnabled()`.
+
+With this new API method the developer could write one method that performs the
+same action for all three, potentially with small differences based on the
+loaded cause.
+
+More use cases can be found [here](https://github.com/w3c/webextensions/issues/353#issuecomment-1582536300).
+
+### Known Consumers
+
+This is a broadly applicable API method that will be helpful to any developer
+wanting to execute code as the extension changes running states. But there
+appears to be strong interest per the [WECG discussion](https://github.com/w3c/webextensions/issues/353).
+
+## Specification
+
+### Schema
+#### `chrome.runtime.onEnabled`
+```typescript
+namespace runtime {
+  // Fired when an extension goes from being in a disabled state to an enabled
+  // state.
+  export interface onEnabled {
+    addListener(callback: () => void): void;
+    removeListener(callback: () => void): void;
+  }
+}
+```
+
+#### `chrome.runtime.onExtensionLoaded`
+```typescript
+namespace runtime {
+  // The reason for which the event is being dispatched.
+  //
+  // 'enabled': The extension was re-enabled from a disabled state.
+  //
+  // 'installed': The extension was newly installed.
+  //
+  // 'updated': The extension was reloaded after an update.
+  //
+  // 'profileStart': The extension is being loaded during a profile startup.
+  //
+  // 'reload': The extension was reloaded (e.g. via `chrome.runtime.reload() or`
+  // the user manually reloaded the extension).
+  export type OnLoadedReason = 'enabled' | 
+                               'installed' |
+                               'updated' |
+                               'profileStart' |
```

Comment:
I don’t think “profile” should be in the name here. I think this should be `started` (to be consistent with the other -ed words). Not all browsers have profiles, and many users only use a single profile, so referencing “profile” could create unnecessary confusion or imply a narrower scope than intended.

All the other reasons — `enabled`, `installed`, `updated`, and `reload` — also apply to profiles since these actions occur within a specific profile context, but they avoid explicitly mentioning profiles.

Similarly, using a name like started ensures consistency with the other past-tense reason names while keeping the terminology universal and clear.

Reply:
Makes sense. I have renamed it to started.

Reply:
I don't think we should use "started", because it's ambiguous what "started" means -- without additional explanation, that could also mean the background context started, or the extension got an active process, or any number of other things.

Maybe "browserStart"?  Technically, that's not _strictly_ true, since in the case of browsers that support profiles, the browser may have already been running -- but given extensions are scoped to profiles, I think this abstraction is okay.

WDYT?

Reply:
@rdcronin `browserStart` also introduces ambiguity. It suggests the event is tied directly to the browser process starting, which as you say, isn’t _strictly_ true, especially with multiple profiles (the opposite problem of the original `profileStart`). This can mislead developers into thinking the event is narrower in scope than it is and does not apply to profiles.

My main concern with “profile” (besides the tense and camel case — we really need to be [consistent with value strings](https://github.com/w3c/webextensions/issues/563#issuecomment-2207298858)) is that this is, as far as I know, the first mention of profiles in the extension API. While the other enum values equally apply to profiles, they don’t mention profiles — and they shouldn’t. (If all enum values equally apply to the same context, explicitly mentioning the context in the values is redundant and unnecessary.) Hence, why I still think `started` is perfectly fine since it applies to the profile just like all the other reasons.

If `started` isn’t acceptable, another option would be `initialized`. It’s clear, concise, and consistent with past-tense naming conventions like `enabled`, `installed`, and `updated`. It has the connotation that it only happens once per profile load, while keeping the terminology flexible enough to apply across different scenarios, such as profile or browser-level startup.

Reply:
@xeenon the reason I'm okay with the ambiguity of `browserStart` is that, from the extension's POV, it *is* the browser start -- no other existing windows or behavior is exposed to the extension, so it shouldn't know or care about them.  There could always be other browser instances running (multiple channels or running completely different browsers), so I don't think any extension should ever assume that, even with "browserStart", it means "there are definitely no other browser-like things running".  That said, if we wanted to avoid browserStart, how about `sessionStart`?  That has its own set of ambiguity ("session" on the web means something different), but is perhaps more clear?

Agreed we need to come to a conclusion on the value strings.  Let's plan on discussing that soon.  I don't think it's a scalable solution to say we'll only add one-word enum values : )

In the interest of parallelizing this work, WDYT about adding a specific note that this name is TBD, but otherwise allowing this to proceed (perhaps merge with an explicit TODO or similar)?  We can be sure to finalize the name before we launch the API in any browser.

Reply:
@rdcronin I'd be fine with `session-started` (or `sessionStarted`, if that is the case style going forward). That fits with the lifetime of `storage.session`. I wouldn't want to resolve to say this is TBD, as that just sets it in stone once one of the browsers ship it — API is permanent.

I think we should write an API style guide to cover things like this for easy referral to by all parties as we write API proposals. I'll file an issue for that.

Reply:
Personally, I like the word `init`, `startup` or `firstStart`. It is the starting point of the extension's lifecycle. We already use the name in `runtime.onStartup()`.

`browserStart` is OK for me. I don't like the word `session` , it's ambiguous for me.

Reply:
@xeenon To clarify, I'm suggesting we unblock implementation work, *not* that we ship this in any browser.  We would absolutely have it be a requirement to finalize all names before shipping anything.  But if the only remaining bit is bikeshedding the name, I think Sohail could get a start on the implementation while we discuss, rather than blocking that.  Does that sound reasonable?  (Though, maybe moot, depending on below)

@hanguokai I like `startup`, which is subtly different IMO from `started`, and as you pointed out, has precedence with `runtime.onStartup`.  I don't think `firstStart` makes sense, because it's ambiguous what "first" is related to ("firstStart" could also mean "install"), and I think "init" is generally unclear.

Anyone else have thoughts on `startup` vs `browser-start` vs `session-start`?

Reply:
I'm fine with `startup`.

Reply:
:tada: Thanks for the suggestion, @hanguokai , and for the discussion, @xeenon !

Reply:
Yes, having `startup` makes sense as it is similar to `runtime.onStartup`. I have updated the proposal. Thank you for the inputs here @rdcronin, @xeenon and @hanguokai 😄